Add Effect Filter

alexsosnovsky opened the pull request:
The new `effect` filter returns `ALLOW` if the player has an effect of the same type as the filter. If a mapmaker wants to be more selective, there are three optional attributes: `min-duration`, `max-duration`, and `amplifier`.

Syntax example:
```xml
<filters>
    <effect id="filter1" min-duration="10s" amplifier="1">saturation</effect>
    <effect id="filter2">saturation</effect>
    <effect id="filter3" max-duration="30s" amplifier="2">night vision</effect>
    <effect id="filter4" amplifier="1">water breathing</effect>
    <effect id="filter5" min-duration="10s" max-duration="20s" amplifier="2">night vision</effect>
</filters>
```

alexsosnovsky committed:
Add effect filter

Signed-off-by: alexsosnovsky <60436249+alexsosnovsky@users.noreply.github.com>

Pablete1234 commented:
I do not see the point in having a filter with effects as children, i think no other filter does this, instead, it's the filter itself that has the properties. 
Also, there's no use for has-*, for example team filter is simply team, even if the name itself is reused elsewhere
```xml
<filters>
    <effect id="filter1" duration="0s" amplifier="4">saturation</effect>
    <effect id="filter2" duration="30s" amplifier="1">water breathing</effect>
    <effect id="filter3" duration="0s" amplifier="2">night vision</effect>
</filters>
```

Regarding duration, a filter with a *specific* duration is of little use, it should probably be a range: `duration="2s..15s"` or `duration="2s.."`

Pugzy commented:
Not sure if these exists in this version of PGM but it looks like ranges and comparisons were implemented like so https://pgm.dev/docs/modules/mechanics/filters/#random-filter which may be worth replicating as the syntax used in the comment above.

alexsosnovsky committed:
Revamp attributes

Signed-off-by: alexsosnovsky <60436249+alexsosnovsky@users.noreply.github.com>

alexsosnovsky commented:
> I do not see the point in having a filter with effects as children, i think no other filter does this, instead, it's the filter itself that has the properties.
> Also, there's no use for has-*, for example team filter is simply team, even if the name itself is reused elsewhere
> Regarding duration, a filter with a _specific_ duration is of little use, it should probably be a range: `duration="2s..15s"` or `duration="2s.."`

I've updated it to simply use the `<effect/>` tag. I also added optional `min-duration` and `max-duration` attributes so that a range can be created easily. The sample XML in the original post has been updated.


> Not sure if these exists in this version of PGM but it looks like ranges and comparisons were implemented like so https://pgm.dev/docs/modules/mechanics/filters/#random-filter which may be worth replicating as the syntax used in the comment above.

This looks interesting but I felt it was simpler and more intuitive to just have `min-duration` and `max-duration` attributes which take normal duration strings as input.

alexsosnovsky committed:
Merge branch 'dev' into effect

alexsosnovsky committed:
Merge branch 'dev' into effect

Pablete1234 commented on core/src/main/java/tc/oc/pgm/filters/FilterParser.java:
```diff
@@ -470,6 +470,19 @@ public WearingItemFilter parseWearingItem(Element el) throws InvalidXMLException
 
+  @MethodParser("effect")
+  public HasEffectFilter parseHasEffect(Element el) throws InvalidXMLException {
+    Attribute minAttribute = el.getAttribute("min-duration");
```
Make use of PGM nodes  instead of XML attributes, eg:: `XMLUtils.parseDuration(Node.fromAttr(el, "min-duration"))`

alexsosnovsky committed:
Use nodes

Signed-off-by: alexsosnovsky <60436249+alexsosnovsky@users.noreply.github.com>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/filters/HasEffectFilter.java:
```diff
@@ -0,0 +1,42 @@
+import tc.oc.pgm.api.filter.query.PlayerQuery;
+import tc.oc.pgm.api.player.MatchPlayer;
+
+public class HasEffectFilter extends ParticipantFilter {
```
```suggestion
public class EffectFilter extends ParticipantFilter {
```

Pablete1234 commented on core/src/main/java/tc/oc/pgm/filters/HasEffectFilter.java:
```diff
@@ -0,0 +1,42 @@
+public class HasEffectFilter extends ParticipantFilter {
+  protected final PotionEffect base;
+  // min/max duration are stored in ticks
+  protected final long minDuration;
+  protected final long maxDuration;
```
Use an actual `Range<Long>`, pass it as part of the constructor

Pablete1234 commented on core/src/main/java/tc/oc/pgm/filters/HasEffectFilter.java:
```diff
@@ -0,0 +1,42 @@
+
+      if (effect.getType().equals(base.getType())
+          && (!amplifier || (effect.getAmplifier() == base.getAmplifier()))
+          && ((minDuration == -1) || (effect.getDuration() >= minDuration))
+          && ((maxDuration == -1) || (effect.getDuration() <= maxDuration))) {
```
This logic will be greatly simplified with the range, `range.contains(effect.getDuration())`

Pablete1234 commented on core/src/main/java/tc/oc/pgm/filters/FilterParser.java:
```diff
@@ -470,6 +471,18 @@ public WearingItemFilter parseWearingItem(Element el) throws InvalidXMLException
 
+  @MethodParser("effect")
+  public HasEffectFilter parseHasEffect(Element el) throws InvalidXMLException {
+    Duration minDuration = XMLUtils.parseDuration(Node.fromAttr(el, "min-duration"));
+    Duration maxDuration = XMLUtils.parseDuration(Node.fromAttr(el, "max-duration"));
```
Create a Range.atMost, Range.atLeast, or Range.closed based on if the durations are present or not, and use that in the filter

Pablete1234 commented on core/src/main/java/tc/oc/pgm/filters/FilterParser.java:
```diff
@@ -470,6 +471,18 @@ public WearingItemFilter parseWearingItem(Element el) throws InvalidXMLException
+  public HasEffectFilter parseHasEffect(Element el) throws InvalidXMLException {
+    Duration minDuration = XMLUtils.parseDuration(Node.fromAttr(el, "min-duration"));
+    Duration maxDuration = XMLUtils.parseDuration(Node.fromAttr(el, "max-duration"));
+    boolean amplifier = Node.fromAttr(el, "amplifier") != null ? true : false;
```
```suggestion
    boolean amplifier = Node.fromAttr(el, "amplifier") != null;
```

alexsosnovsky committed:
Use range

Signed-off-by: alexsosnovsky <60436249+alexsosnovsky@users.noreply.github.com>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/filters/EffectFilter.java:
```diff
@@ -31,8 +30,7 @@ protected QueryResponse queryPlayer(PlayerQuery query, MatchPlayer player) {
           && (!amplifier || (effect.getAmplifier() == base.getAmplifier()))
-          && ((minDuration == -1) || (effect.getDuration() >= minDuration))
-          && ((maxDuration == -1) || (effect.getDuration() <= maxDuration))) {
+          && duration.contains((long) effect.getDuration())) {
```
I was under the impression that getDuration was a long, seems like it's an int, so the Range should also be a `Range<Integer>`, and you should do the cast in the parser instead, but it's a somewhat minor issue

alexsosnovsky committed:
Cast in parser

Signed-off-by: alexsosnovsky <60436249+alexsosnovsky@users.noreply.github.com>

alexsosnovsky committed:
Merge branch 'dev' into effect

alexsosnovsky committed:
Merge branch 'dev' into effect

alexsosnovsky committed:
Merge branch 'dev' into effect
